projects/interfaces/filtering/filtering.go

```diff
@@ -0,0 +1,28 @@
+package filtering
+
+import (
+	"io"
+)
+
+type filteringPipe struct {
+	fw io.Writer
+}
+
+func NewFilteringPipe(w io.Writer) io.Writer {
+	return filteringPipe{fw: w}
+}
+
+func (f filteringPipe) Write(p []byte) (n int, err error) {
+	filtered := []byte{}
+	for _, value := range p {
+		if value < '0' || value > '9' {
+			filtered = append(filtered, value)
+		}
+	}
```

Comment:
This works, and I think is a good implementation, but I have some questions about trade-offs!

1. What's the difference between making this intermediate byte slice to call Write once, vs calling Write each byte at a time?
2. You're currently unconditionally making a byte-slice and copying the bytes to it - potentially all of the bytes. Are there times you don't need to do this? Can you think of how you could _always_ avoid copying? What would the trade-offs be of each of these approaches?
3. If you _are_ going to unconditionally make a byte slice and copy all of the non-number bytes into it, you may want to pre-allocate the capacity (e.g. with something like `filtered := make([]byte, 0, len(p))`, or even something more fancy) rather than what you're currently doing. Can you work out and talk through what the differences are?

Reply:
> 1.  What's the difference between making this intermediate byte slice to call Write once, vs calling Write each byte at a time?

**Creating intermediate slice :**  We call write function once , and this particularly efficient if we have a large input , or if we filter infrequently ( situation when we have numbers more frequent than other characters)

**Calling Write for each bytes  :** Can be more efficient if we are filtering frequently or if we have a small input.

> 2. You're currently unconditionally making a byte-slice and copying the bytes to it - potentially all of the bytes.  

> Are there times you don't need to do this?

we don't need to copy if the slice doesn't includes any number, so we can return the original one (found method `contains` on `slices` packages we can use it to check the slice doesn't includes numbers . Actually I found `DeleteFunc` which does the filtering , it doesn't  copy till it found an element to delete:) ).

> Can you think of how you could always avoid copying? What would the trade-offs be of each of these approaches?

Yes we can change the original slice directly by removing the numbers , (kind of splitting the slice and joining it again) but I think it's inefficient as we may end up moving all the elements.
|| Intermediate slice | Remove |
|----|---|----|
|Pros|Preserve integrity of the original slice| Avoid memory allocation and copying|
|Cons|memory usage and copying|Doesn't preserve original slice, shifting elements|

> 3. If you are going to unconditionally make a byte slice and copy all of the non-number bytes into it, you may want to pre-allocate the capacity (e.g. with something like filtered := make([]byte, 0, len(p)), or even something more fancy) rather than what you're currently doing. Can you work out and talk through what the differences are?

unconditionally make a byte slice > exceed the capacity of the slice > resizing the slice by increasing capacity > allocating new underlying array > allocating the array in the memory > copying elements from the old array to the new one. 
So we are frequently resizing the slice,  allocating memory and copying  :/
by pre-allocate the capacity we preserve the same underlying array as the max length that the slice can have is `len(p)` (so we are not exceeding its capacity > no memory allocation and no resizing) ,  .

Reply:
Generally good thoughts!

One more thing to consider about the few-writes vs multiple-writes is it actually depends on the behaviour of the writer you're calling. Some writers are cheap to write to (and maybe even internally do buffering to make sure they're cheap to write to), and others are expensive to write to. Imagine a hypothetical writer where every time you call write, it sends those bytes across the internet and waits until it hears back that they've been received. Each call to `Write` may then take hundreds or milliseconds or more! On the other hand, if the writer you're writing to just adds the bytes you call it with to a buffer until you close the writer, or maybe call something like [`Flush`](https://pkg.go.dev/bufio#Writer.Flush), or until a certain number of bytes were written, or until no more bytes have been written for 1ms, or something, then the performance of a single large write vs multiple small writes will be roughly the same.

Unfortunately, you don't know from just having an `io.Writer` how fast/slow the writer is going to be. What do you think that means in terms of how you should make choices in your code?

> Yes we can change the original slice directly by removing the numbers

Unfortunately not! As per [the `Writer` docs](https://pkg.go.dev/io#Writer):

> Write must not modify the slice data, even temporarily.

This means we either need to make reads work in-place, or we need to copy to some intermediate memory.

A few things to consider/contrast in terms of these behaviours:

1. As you described, pre-allocating can avoid needing to repeatedly grow+copy a buffer allocation (but note that, particularly for small-ish buffers, Go probably behaves pretty well, so this is probably a small-ish optimisation - worth benchmarking, though!)
2. Taking sub-slice views over slices is cheap (remember a slice is just a pointer to a buffer + length + capacity, so making a new slice that shares the same underlying buffer is cheap) - so you could write some code along the lines of:

```go
start := 0
for i, byte := range p {
    if isNumber(byte) {
        f.fw.Write(p[start:i]) // TODO: Handle errors and stuff, and maybe skip the write if the slice is empty
        start = i + 1
    }
}
if start < len(p) {
    f.fw.Write(p[start:])  // TODO: Handle errors and stuff
}
```

(This has that "may end up doing many writes" problem, but avoids _any_ copying - as with many things, neither is _obviously_ better, but there are trade-offs between "never copy" and "minimise writes" :))